source/PluginDev/Assets/GooglePlayGames/Platforms/Android/AndroidRealTimeMultiplayerClient.cs

```diff
@@ -34,111 +54,736 @@ public AndroidRealTimeMultiplayerClient(AndroidJavaObject account)
+                            {
+                                AndroidTaskUtils.AddOnFailureListener(
+                                    task,
+                                    e =>
+                                    {
+                                        listener.OnRoomConnected(false);
+                                        CleanSession();
+                                    });
+                            }
+                        }
+                    });
+            }
         }
 
         public void SendMessageToAll(bool reliable, byte[] data)
         {
-            // Task<Void> sendUnreliableMessageToOthers(@NonNull byte[] messageData, @NonNull String roomId)
+            if (reliable)
+            {
+                List<Participant> participants = GetConnectedParticipants();
+                foreach (Participant participant in participants)
+                {
+                    SendMessage(true, participant.ParticipantId, data);
+                }
+
+                return;
+            }
+
+            int roomStatus = GetRoomStatus();
+            if (roomStatus != ROOM_STATUS_ACTIVE && roomStatus != ROOM_STATUS_CONNECTING)
+            {
+                OurUtils.Logger.d("Sending message is not allowed in this state.");
+                return;
+            }
+
+            string roomId = mRoom.Call<string>("getRoomId");
+            mClient.Call<AndroidJavaObject>("sendUnreliableMessageToOthers", data, roomId);
         }
 
         public void SendMessageToAll(bool reliable, byte[] data, int offset, int length)
         {
-            // Task<Void> sendUnreliableMessageToOthers(@NonNull byte[] messageData, @NonNull String roomId)
+            SendMessageToAll(reliable, Misc.GetSubsetBytes(data, offset, length));
         }
 
         public void SendMessage(bool reliable, string participantId, byte[] data)
         {
-            // Task<Integer> sendReliableMessage(@NonNull byte[] messageData, @NonNull String roomId, @NonNull String recipientParticipantId, @Nullable ReliableMessageSentCallback callback)
-            // Task<Void> sendUnreliableMessage(@NonNull byte[] messageData, @NonNull String roomId, @NonNull String recipientParticipantId)
+            int roomStatus = GetRoomStatus();
+            if (roomStatus != ROOM_STATUS_ACTIVE && roomStatus != ROOM_STATUS_CONNECTING)
+            {
+                OurUtils.Logger.d("Sending message is not allowed in this state.");
+                return;
+            }
+
+            if (GetParticipant(participantId) == null)
+            {
+                OurUtils.Logger.e("Attempted to send message to unknown participant " + participantId);
+                return;
+            }
+
+            string roomId = mRoom.Call<string>("getRoomId");
+            if (reliable)
+            {
+                using (var callback = new AndroidJavaObject("com.google.games.bridge.ReliableMessageSentCallbackProxy",
```

Comment:
callback is not null?

Reply:
Removed proxy/bridge class relating to this callback as they're not used.